### PR TITLE
Add support for the list branched endpoint (47deg#231)

### PR DESCRIPTION
**Pull request by YarekTyshchenko**
- [x] Create tests for protected query parameter
- [ ] Fix build
- [ ] rebase / squash as needed

### CODE REVIEW

File: github4s/shared/src/main/scala/github4s/free/algebra/RepositoryOps.scala [hunk 1]
```diff
@@ -170,6 +177,15 @@ class RepositoryOps[F[_]](implicit I: InjectK[RepositoryOp, F]) {
     Free.inject[RepositoryOp, F](
       ListCommits(owner, repo, sha, path, author, since, until, pagination, accessToken))
 
+  def listBranches(
+      owner: String,
+      repo: String,
+      `protected`: Option[Boolean] = None,
+      accessToken: Option[String] = None
+  ): Free[F, GHResponse[List[Branch]]] =
+    Free.inject[RepositoryOps, F](
```
Comment (JesusMtnez) [inline github4s/shared/src/main/scala/github4s/free/algebra/RepositoryOps.scala]:
```suggestion
    Free.inject[RepositoryOp, F](
```

**Comment by JesusMtnez** [comment]
@YarekTyshchenko Apart from the minor typo, the decoders for the new entities (`Branch` and `BranchCommit`) are missing and tests are failing.

It should not be hard to add, but let me know if you need help doing so.

File: github4s/shared/src/main/scala/github4s/free/algebra/RepositoryOps.scala (hunk 1 above)
Comment (YarekTyshchenko) [reply github4s/shared/src/main/scala/github4s/free/algebra/RepositoryOps.scala]: Well spotted

File: github4s/shared/src/main/scala/github4s/Decoders.scala [hunk 2]
```diff
@@ -50,6 +50,32 @@ object Decoders {
       )
   }
 
+  implicit val decodeBranch: Decoder[Branch] = Decoder.instance { c ⇒
+    for {
+      name            ← c.downField("name").as[String]
+      commit          ← c.downField("commit").as[BranchCommit]
+      branchProtected ← c.downField("protected").as[Boolean]
```
Comment (JesusMtnez) [inline github4s/shared/src/main/scala/github4s/Decoders.scala]:
```suggestion
      branchProtected ← c.downField("protected").as[Option[Boolean]]
```

File: github4s/shared/src/main/scala/github4s/Decoders.scala [hunk 3]
```diff
@@ -50,6 +50,32 @@ object Decoders {
...
+      protection_url  ← c.downField("protection_url").as[String]
```
Comment (JesusMtnez) [inline github4s/shared/src/main/scala/github4s/Decoders.scala]:
```suggestion
      protection_url  ← c.downField("protection_url").as[Option[String]]
```

File: github4s/shared/src/main/scala/github4s/free/domain/Repository.scala [hunk 4]
```diff
@@ -103,6 +103,18 @@ case class Commit(
     author_url: Option[String]
 )
 
+case class Branch(
+  name: String,
+  commit: BranchCommit,
+  `protected`: Boolean,
```
Comment (JesusMtnez) [inline github4s/shared/src/main/scala/github4s/free/domain/Repository.scala]:
```suggestion
  `protected`: Option[Boolean],
```

File: github4s/shared/src/main/scala/github4s/free/domain/Repository.scala [hunk 5]
```diff
@@ -103,6 +103,18 @@ case class Commit(
...
+  protection_url: String
```
Comment (JesusMtnez) [inline github4s/shared/src/main/scala/github4s/free/domain/Repository.scala]:
```suggestion
  protection_url: Option[String]
```

File: github4s/shared/src/test/scala/github4s/utils/TestData.scala [hunk 6]
```diff
@@ -360,6 +360,16 @@ trait TestData extends DummyGithubUrls {
...
     None
   )
+  val validBranchName = "master"
+  val branch = Branch(
+    name = validBranchName,
+    commit = BranchCommit(
+      sha = validCommitSha,
+      url = s"https://api.github.com/repos/$validRepoOwner/$validRepoName/commits/$validCommitSha"
+    ),
+    `protected` = true,
```
Comment (JesusMtnez) [inline github4s/shared/src/test/scala/github4s/utils/TestData.scala]:
```suggestion
    `protected` = Some(true),
```

File: github4s/shared/src/test/scala/github4s/utils/TestData.scala [hunk 7]
```diff
@@ -360,6 +360,16 @@ trait TestData extends DummyGithubUrls {
...
+    protection_url = s"https://api.github.com/repos/$validRepoOwner/$validRepoName/branches/$validBranchName/protection"
```
Comment (JesusMtnez) [inline github4s/shared/src/test/scala/github4s/utils/TestData.scala]:
```suggestion
    protection_url = Some(s"https://api.github.com/repos/$validRepoOwner/$validRepoName/branches/$validBranchName/protection")
```

**Comment by YarekTyshchenko** [comment]
@JesusMtnez You are right, I must have missed that bit in the docs.`Option` sounds good!

**Comment by fedefernandez** [comment]
hey @YarekTyshchenko, the tests use a GitHub token that is not available in forks. You'd need to create a Personal Access token in Github with `gists` permissions (I don't remember if anyone more is necessary) and store it in the environment variable `GITHUB4S_ACCESS_TOKEN` before executing the tests. I can also execute them locally, I just need to find some time. Thanks

**Comment by JesusMtnez** [comment]
@YarekTyshchenko Greenlight! All checks are passing, we are good to merge.

**Comment by JesusMtnez** [comment]
@YarekTyshchenko do you need a new release of `github4s`? Just let me know if needed :wink:

**Comment by YarekTyshchenko** [comment]
@JesusMtnez Let me have a quick look at #233 first. I think that should be an easy one

**Pull request merged**